src/logic/player.cc

```diff
@@ -443,6 +443,10 @@ MessageId Player::add_message(Game& game, std::unique_ptr<Message> new_message,
 	if (message->serial() > 0) {
 		MapObject* mo = egbase().objects().get_object(message->serial());
 		mo->removed.connect([this, id](unsigned) { message_object_removed(id); });
+		upcast(Building, site, mo);
+		if (site) {
+			site->muted.connect([this, id](unsigned) { message_object_removed(id); });
+		}
```

Comment:
`upcast` has bad performance because it uses `dynamic_cast` internally, it is more [efficient](https://stackoverflow.com/questions/9959315/why-virtual-function-call-is-faster-than-dynamic-cast) to check the `MapObjectType` first:
```suggestion
		if (mo->descr().type() >= MapObjectType::BUILDING) {
			upcast(Building, site, mo);
			assert(site != nullptr);
			site->muted.connect([this, id](unsigned) { message_object_removed(id); });
		}
```

Reply:
thx, done